### PR TITLE
fix: don't re-download already bundled pmd on installation

### PR DESCRIPTION
**Pull request by YodaDaCoda**
PMD was accidentally re-downloaded when installing the `pmd-bin` package although it is already bundled in this package.

We only need to download an OS-specific version of `java`. `PMD` itself is not OS-specific.

This change also removes some `dependencies` again, which are only used to build this package.

### CODE REVIEW

**Comment by amtrack** [comment]
@YodaDaCoda Hmm, I think I'd prefer to bundle pmd again in the package like it was before.

- pmd: download using the `prepare` hook so that it is included in this npm package (OS independent)
- java: dynamically downloaded on installation (OS dependent)

What do you think?

**Comment by YodaDaCoda** [comment]
PMD is still bundled with the package, I'm pretty sure it gets re-downloaded on the end-user's machine however. Obviously not ideal.

That's a fantastic idea. So we'd have a `prepare.js` which would download pmd, and an `install.js` which would prepare java.

I'll update the PMD downloading to rename the directory to just pmd-bin to match the previous structure.

**Comment by YodaDaCoda** [comment]
After reading through the docs, I think we want to use `prepublishOnly` not `prepare`, as `prepare` is run as part of `npm install`.

https://docs.npmjs.com/cli/v7/using-npm/scripts

**Comment by YodaDaCoda** [comment]
@amtrack Looks like the CI environment probably needs updating to handle this, since we're no longer downloading PMD on the end-user's machine and the CI environment is running `npm ci` from within the repo, PMD doesn't get downloaded in the CI env. This wouldn't be an issue on an end-user's machine as PMD would be included/distributed from npm.

Not sure how you'd like to handle that. You could run `npm run prepublishOnly` within the CI, but that doesn't seem to me the proper way of doing things as it's not 100% replicating that which would be executed on an end-user's machine.

**Comment by amtrack** [comment]
@YodaDaCoda Thank you so much!

I've pushed a commit on your branch to fix the generated `package-lock.json` to fully declare them as `devDependencies`.

Regarding the `prepare` vs `prepublishOnly` lifecycle scripts:

I've always preferred the `prepare` script because it makes **building** the npm package more easy.

`npm install` in CI would build (prepare) the package so that it is ready to be used/tested.

From the docs about the `prepare` script:

> Runs on local npm install **without any arguments**

The way I understand this, is that it is only executed when we run `npm install` **in the pmd-bin repository**.
As far as I know, the `prepare` script isn't called when a user installs a published version of pmd-bin from an npm registry using `npm install pmd-bin`.
I've tested it looking at the logs of `npm install pmd-bin@1.19.0 --verbose`.

The prepare script is also quite handy, when we want to install pmd-bin e.g. from a fork like yours using `npm install git+https://github.com/YodaDaCoda/pmd-bin.git`.
In this case npm clones the git repository and runs `npm install` in the cloned pmd-bin repo, which will build the package so that it is ready to be used/tested.

**In any case**, to make things more clear, we could rename the `prepare.js` file to s.th. like `download-pmd.js`, create a `build` script for it and call the build script from the lifecycle script explicitly:

```json
"build": "node download-pmd.js",
"prepare": "npm run build",
```

or if we use `prepublishOnly`:

```json
"build": "node download-pmd.js",
"prepublishOnly": "npm run build",
```

In the latter case, we would need to run `npm install && npm run build` in CI, which reads much better than `npm install && npm run prepublishOnly` in my opinion.

**Comment by YodaDaCoda** [comment]
> The way I understand this, is that it is only executed when we run `npm install` in the pmd-bin repository.
As far as I know, the `prepare` script isn't called when a user installs a published version of pmd-bin from an npm registry using `npm install pmd-bin`.

If this is indeed the case (it sounds like you know better than I) then `prepare` would be the way to go IMO. Thanks for the explanation. I'm not as familiar with some of the intricacies of NPM as I'd like.

I certainly prefer `prepare` over `prepublishOnly` for readability and I think having a build script is a great idea. I'll add a commit to Do The Thing tomorrow.

File: .github/workflows/default.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ jobs:
         with:
           node-version: 14
       - name: Install dependencies
-        run: npm ci
+        run: npm install && npm run build
```
Comment (amtrack) [inline .github/workflows/default.yml:22]: `npm ci` (clean install) should have already run the `prepare` lifecycle script, so we don't need to run the build script again.

```suggestion
        run: npm ci
```

File: .github/workflows/default.yml [hunk 2]
```diff
@@ -37,7 +37,7 @@ jobs:
           unset JAVA_HOME
           java --version || true
       - name: Install dependencies
-        run: npm ci
+        run: npm install && npm run build
```
Comment (amtrack) [inline .github/workflows/default.yml:40]:
```suggestion
        run: npm ci
```

**Review by amtrack — CHANGES_REQUESTED** [review]
I have requested some minor changes for CI.
Besides that, this looks great and I'm more than happy to merge it!

**Pull request merged**